src/cryptography/x509.py

```diff
@@ -1594,3 +1596,155 @@ def sign(self, private_key, algorithm, backend):
+        """
+        if not isinstance(number, six.integer_types):
+            raise TypeError('Serial number must be of integral type.')
+        if self._serial_number is not None:
+            raise ValueError('The serial number may only be set once.')
+        return CertificateBuilder(
+            self._version, self._issuer_name, self._subject_name,
+            self._public_key, number, self._not_valid_before,
+            self._not_valid_after, self._extensions
+        )
+
+    def not_valid_before(self, time):
+        """
+        Sets the certificate activation time.
+        """
+        # TODO: require UTC datetime?
+        if not isinstance(time, datetime.datetime):
+            raise TypeError('Expecting datetime object.')
+        if self._not_valid_before is not None:
+            raise ValueError('The not valid before may only be set once.')
+        return CertificateBuilder(
+            self._version, self._issuer_name, self._subject_name,
+            self._public_key, self._serial_number, time,
+            self._not_valid_after, self._extensions
+        )
+
+    def not_valid_after(self, time):
+        """
+        Sets the certificate expiration time.
+        """
+        # TODO: require UTC datetime?
+        if not isinstance(time, datetime.datetime):
+            raise TypeError('Expecting datetime object.')
+        if self._not_valid_after is not None:
+            raise ValueError('The not valid after may only be set once.')
+        return CertificateBuilder(
+            self._version, self._issuer_name, self._subject_name,
+            self._public_key, self._serial_number, self._not_valid_before,
+            time, self._extensions
+        )
+
+    def add_extension(self, extension, critical):
+        """
+        Adds an X.509 extension to the certificate.
+        """
+        if isinstance(extension, BasicConstraints):
+            extension = Extension(OID_BASIC_CONSTRAINTS, critical, extension)
+        elif isinstance(extension, SubjectAlternativeName):
+            extension = Extension(
+                OID_SUBJECT_ALTERNATIVE_NAME, critical, extension
+            )
+        else:
+            raise NotImplementedError('Unsupported X.509 extension.')
+
+        # TODO: This is quadratic in the number of extensions
+        for e in self._extensions:
+            if e.oid == extension.oid:
+                raise ValueError('This extension has already been set.')
+
+        return CertificateBuilder(
+            self._version, self._issuer_name, self._subject_name,
+            self._public_key, self._serial_number, self._not_valid_before,
+            self._not_valid_after, self._extensions + [extension]
+        )
+
+    def sign(self, backend, private_key, algorithm):
+        """
+        Signs the certificate using the CA's private key.
+        """
+        builder = self
+        if self._version is None:
+            builder = self.version(Version.v3)
```

Comment:
Writing tests for an X.509 library.

On Thu, Jul 23, 2015 at 1:45 PM, Will Bond <notifications@github.com> wrote:

> In src/cryptography/x509.py
> <https://github.com/pyca/cryptography/pull/2098#discussion_r35349775>:
>
> > +            if e.oid == extension.oid:
> > +                raise ValueError('This extension has already been set.')
> > +
> > +        return CertificateBuilder(
> > +            self._version, self._issuer_name, self._subject_name,
> > +            self._public_key, self._serial_number, self._not_valid_before,
> > +            self._not_valid_after, self._extensions + [extension]
> > +        )
> > +
> > +    def sign(self, backend, private_key, algorithm):
> > +        """
> > +        Signs the certificate using the CA's private key.
> > +        """
> > +        builder = self
> > +        if self._version is None:
> > +            builder = self.version(Version.v3)
>
> @reaperhulk <https://github.com/reaperhulk> Not questioning your
> reasoning, just looking to learn – in what situations is a v1 cert valuable?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2098/files#r35349775>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084